drivers/wifi/esp32/src/esp_wifi_drv.c

```diff
@@ -307,20 +307,84 @@ static void esp_wifi_handle_sta_connect_event(void *event_data)
 #endif
 }
 
+static void handle_disconnect_event_while_not_connected(wifi_event_sta_disconnected_t *event)
+{
+	switch (event->reason) {
+	case WIFI_REASON_ROAMING:
+		break;
+	case WIFI_REASON_ASSOC_LEAVE:
+		LOG_DBG("Disconnect Requested");
+		break;
+	case WIFI_REASON_4WAY_HANDSHAKE_TIMEOUT:
+	case WIFI_REASON_HANDSHAKE_TIMEOUT:
+		LOG_DBG("STA Auth Error");
+		wifi_mgmt_raise_connect_result_event(esp32_wifi_iface,
+						     WIFI_STATUS_CONN_WRONG_PASSWORD);
+		break;
+	case WIFI_REASON_DISCONN_INACTIVITY:
+	case WIFI_REASON_TIMEOUT:
+		LOG_DBG("STA Connection Timeout");
+		wifi_mgmt_raise_connect_result_event(esp32_wifi_iface, WIFI_STATUS_CONN_TIMEOUT);
+		break;
+	case WIFI_REASON_NO_AP_FOUND:
+		LOG_DBG("AP Not found");
+		wifi_mgmt_raise_connect_result_event(esp32_wifi_iface,
+						     WIFI_STATUS_CONN_AP_NOT_FOUND);
+		break;
+	default:
+		LOG_DBG("Generic Failure");
+		wifi_mgmt_raise_connect_result_event(esp32_wifi_iface, WIFI_STATUS_CONN_FAIL);
+		break;
+	}
+}
+
+static void handle_disconnect_event_while_connected(wifi_event_sta_disconnected_t *event)
+{
+#if defined(CONFIG_ESP32_WIFI_STA_AUTO_DHCPV4)
+	net_dhcpv4_stop(esp32_wifi_iface);
+#endif
+
+	switch (event->reason) {
+	case WIFI_REASON_ROAMING:
+		LOG_DBG("Roaming");
+		break;
+	case WIFI_REASON_AUTH_LEAVE:
+		LOG_DBG("AP Requested Disconnect");
+		wifi_mgmt_raise_disconnect_result_event(esp32_wifi_iface,
+							WIFI_REASON_DISCONN_AP_LEAVING);
+		break;
+	case WIFI_REASON_ASSOC_LEAVE:
+		LOG_DBG("Disconnect Was Requested");
+		wifi_mgmt_raise_disconnect_result_event(esp32_wifi_iface,
+							WIFI_REASON_DISCONN_USER_REQUEST);
+		break;
+	case WIFI_REASON_AUTH_EXPIRE:
+		LOG_DBG("AP not active");
+		wifi_mgmt_raise_disconnect_result_event(esp32_wifi_iface,
+							WIFI_REASON_DISCONN_INACTIVITY);
+		break;
+	default:
+		LOG_DBG("Generic Failure");
+		wifi_mgmt_raise_connect_result_event(esp32_wifi_iface,
+						     WIFI_REASON_DISCONN_UNSPECIFIED);
+		break;
+	}
+}
+
 static void esp_wifi_handle_sta_disconnect_event(void *event_data)
 {
 	wifi_event_sta_disconnected_t *event = (wifi_event_sta_disconnected_t *)event_data;
 
+	LOG_DBG("Disconnect reason: %d", event->reason);
+
 	if (esp32_data.state == ESP32_STA_CONNECTED) {
-#if defined(CONFIG_ESP32_WIFI_STA_AUTO_DHCPV4)
-		net_dhcpv4_stop(esp32_wifi_iface);
-#endif
-		wifi_mgmt_raise_disconnect_result_event(esp32_wifi_iface, 0);
+		handle_disconnect_event_while_connected(event);
 	} else {
-		wifi_mgmt_raise_disconnect_result_event(esp32_wifi_iface, -1);
+		handle_disconnect_event_while_not_connected(event);
 	}
 
-	LOG_DBG("Disconnect reason: %d", event->reason);
+	wifi_mgmt_raise_disconnect_result_event(esp32_wifi_iface, WIFI_REASON_DISCONN_SUCCESS);
```

Comment:
Perhaps you could return the disconnect event value from both calls you created and have this single call here with that value.
In current approach:
1) `wifi_mgmt_raise_disconnect_result_event` is being called 2x.
2) You have multiple calls to `wifi_mgmt_raise_disconnect_result_event` when you could use a single one.

Would you mind analyzing if the above suggestion is acceptable?